# Ensemble Learning

## Initial Imports

In [27]:
import warnings
warnings.filterwarnings('ignore')

In [28]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

## Read the CSV and Perform Basic Data Cleaning

In [29]:
# Load the data
file_path = Path('Resources/LoanStats_2019Q1.csv')
loan_df = pd.read_csv(file_path)

# Preview the data
loan_df.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    10500.0    0.1719       375.35           RENT     66000.0   
1    25000.0    0.2000       929.09       MORTGAGE    105000.0   
2    20000.0    0.2000       529.88       MORTGAGE     56000.0   
3    10000.0    0.1640       353.55           RENT     92000.0   
4    22000.0    0.1474       520.39       MORTGAGE     52000.0   

  verification_status   issue_d loan_status pymnt_plan    dti  ...  \
0     Source Verified  Mar-2019    low_risk          n  27.24  ...   
1            Verified  Mar-2019    low_risk          n  20.23  ...   
2            Verified  Mar-2019    low_risk          n  24.26  ...   
3            Verified  Mar-2019    low_risk          n  31.44  ...   
4        Not Verified  Mar-2019    low_risk          n  18.76  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim  total_bal_ex_mort total_bc_limit  \
0          65687.0            38199.0         2000.0   
1         271427.0            60641.0        41200.0   
2          60644.0            45684.0         7500.0   
3          99506.0            68784.0        19700.0   
4         219750.0            25919.0        27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              N                     N  
1                     49197.0              N                     N  
2                     43144.0              N                     N  
3                     76506.0              N                     N  
4                     20000.0              N                     N  

[5 rows x 86 columns]

In [30]:
loan_df['home_ownership'] = loan_df['home_ownership'].replace(['OWN'], 0)
loan_df['home_ownership'] = loan_df['home_ownership'].replace(['MORTGAGE'], 1)
loan_df['home_ownership'] = loan_df['home_ownership'].replace(['RENT'], 2)
loan_df['home_ownership'] = loan_df['home_ownership'].replace(['ANY'], 3)
loan_df['verification_status'] = loan_df['verification_status'].replace(['Verified'], 0)
loan_df['verification_status'] = loan_df['verification_status'].replace(['Source Verified'], 1)
loan_df['verification_status'] = loan_df['verification_status'].replace(['Not Verified'], 2)
loan_df['issue_d'] = loan_df['issue_d'].replace(['Jan-2019'], 1.2019)
loan_df['issue_d'] = loan_df['issue_d'].replace(['Feb-2019'], 2.2019)
loan_df['issue_d'] = loan_df['issue_d'].replace(['Mar-2019'], 3.2019)
loan_df['pymnt_plan'] = loan_df['pymnt_plan'].replace(['n'], 0)
loan_df['initial_list_status'] = loan_df['initial_list_status'].replace(['w'], 0)
loan_df['initial_list_status'] = loan_df['initial_list_status'].replace(['f'], 1)
loan_df['next_pymnt_d'] = loan_df['next_pymnt_d'].replace(['Apr-2019'], 4.2019)
loan_df['next_pymnt_d'] = loan_df['next_pymnt_d'].replace(['May-2019'], 5.2019)
loan_df['application_type'] = loan_df['application_type'].replace(['Individual'], 0)
loan_df['application_type'] = loan_df['application_type'].replace(['Joint App'], 1)
loan_df['hardship_flag'] = loan_df['hardship_flag'].replace(['N'], 0)
loan_df['debt_settlement_flag'] = loan_df['debt_settlement_flag'].replace(['N'], 0)
loan_df.head()

loan_amnt  int_rate  installment  home_ownership  annual_inc  \
0    10500.0    0.1719       375.35               2     66000.0   
1    25000.0    0.2000       929.09               1    105000.0   
2    20000.0    0.2000       529.88               1     56000.0   
3    10000.0    0.1640       353.55               2     92000.0   
4    22000.0    0.1474       520.39               1     52000.0   

   verification_status  issue_d loan_status  pymnt_plan    dti  ...  \
0                    1   3.2019    low_risk           0  27.24  ...   
1                    0   3.2019    low_risk           0  20.23  ...   
2                    0   3.2019    low_risk           0  24.26  ...   
3                    0   3.2019    low_risk           0  31.44  ...   
4                    2   3.2019    low_risk           0  18.76  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
0          65687.0            38199.0          2000.0   
1         271427.0            60641.0         41200.0   
2          60644.0            45684.0          7500.0   
3          99506.0            68784.0         19700.0   
4         219750.0            25919.0         27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              0                     0  
1                     49197.0              0                     0  
2                     43144.0              0                     0  
3                     76506.0              0                     0  
4                     20000.0              0                     0  

[5 rows x 86 columns]

## Split the Data into Training and Testing

In [31]:
# Create features
x_cols = [i for i in loan_df.columns if i not in ("loan_status")]
X = loan_df[x_cols]

# Create target
y = loan_df["loan_status"]

In [32]:
X.describe()

loan_amnt      int_rate   installment  home_ownership    annual_inc  \
count  68817.000000  68817.000000  68817.000000    68817.000000  6.881700e+04   
mean   16677.594562      0.127718    480.652863        1.269483  8.821371e+04   
std    10277.348590      0.048130    288.062432        0.654930  1.155800e+05   
min     1000.000000      0.060000     30.890000        0.000000  4.000000e+01   
25%     9000.000000      0.088100    265.730000        1.000000  5.000000e+04   
50%    15000.000000      0.118000    404.560000        1.000000  7.300000e+04   
75%    24000.000000      0.155700    648.100000        2.000000  1.040000e+05   
max    40000.000000      0.308400   1676.230000        3.000000  8.797500e+06   

       verification_status       issue_d  pymnt_plan           dti  \
count         68817.000000  68817.000000     68817.0  68817.000000   
mean              1.330006      1.928072         0.0     21.778153   
std               0.719105      0.743862         0.0     20.199244   
min               0.000000      1.201900         0.0      0.000000   
25%               1.000000      1.201900         0.0     13.890000   
50%               1.000000      2.201900         0.0     19.760000   
75%               2.000000      2.201900         0.0     26.660000   
max               2.000000      3.201900         0.0    999.000000   

        delinq_2yrs  ...  pct_tl_nvr_dlq  percent_bc_gt_75  \
count  68817.000000  ...    68817.000000      68817.000000   
mean       0.217766  ...       95.057627         30.626217   
std        0.718367  ...        8.326426         33.631463   
min        0.000000  ...       20.000000          0.000000   
25%        0.000000  ...       93.000000          0.000000   
50%        0.000000  ...      100.000000         20.000000   
75%        0.000000  ...      100.000000         50.000000   
max       18.000000  ...      100.000000        100.000000   

       pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  total_bal_ex_mort  \
count          68817.000000    68817.0     6.881700e+04       6.881700e+04   
mean               0.125972        0.0     2.100332e+05       6.133843e+04   
std                0.336732        0.0     1.928088e+05       5.738798e+04   
min                0.000000        0.0     3.600000e+03       2.350000e+02   
25%                0.000000        0.0     6.697700e+04       2.650300e+04   
50%                0.000000        0.0     1.467100e+05       4.535700e+04   
75%                0.000000        0.0     3.036400e+05       7.657000e+04   
max                4.000000        0.0     3.292782e+06       1.295455e+06   

       total_bc_limit  total_il_high_credit_limit  hardship_flag  \
count    68817.000000                6.881700e+04        68817.0   
mean     29734.128558                5.572240e+04            0.0   
std      26795.394232                5.095845e+04            0.0   
min        100.000000                1.270000e+02            0.0   
25%      11600.000000                2.288000e+04            0.0   
50%      22100.000000                4.200000e+04            0.0   
75%      39300.000000                7.249900e+04            0.0   
max     509400.000000                1.426964e+06            0.0   

       debt_settlement_flag  
count               68817.0  
mean                    0.0  
std                     0.0  
min                     0.0  
25%                     0.0  
50%                     0.0  
75%                     0.0  
max                     0.0  

[8 rows x 85 columns]

In [33]:
# Check the balance of target values
y.value_counts()

low_risk     68470
high_risk      347
Name: loan_status, dtype: int64

In [34]:
# Split the X and y into X_train, X_test, y_train, y_test
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

## Data Pre-Processing

Scale the training and testing data using the `StandardScaler` from `sklearn`. Remember that when scaling the data, you only scale the features data (`X_train` and `X_testing`).

In [35]:
# Create the StandardScaler instance
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

In [36]:
# Fit the Standard Scaler with the training data
# When fitting scaling functions, only train on the training dataset
X_scaler = scaler.fit(X_train)

In [37]:
# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble classifier . For each algorithm, be sure to complete the folliowing steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Display the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier only, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests

### Balanced Random Forest Classifier

In [38]:
# Resample the training data with the BalancedRandomForestClassifier
from imblearn.ensemble import BalancedRandomForestClassifier
balanced_random_forest_model = BalancedRandomForestClassifier(n_estimators=100, random_state=78)
balanced_random_forest_model = balanced_random_forest_model.fit(X_train_scaled, y_train)
balanced_random_forest_predictions = balanced_random_forest_model.predict(X_test_scaled)

In [39]:
# Calculated the balanced accuracy score
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
balanced_random_forest_accuracy_score = accuracy_score(y_test, balanced_random_forest_predictions)
balanced_random_forest_accuracy_score

0.8643998837547224

In [40]:
# Display the confusion matrix
balanced_random_forest_confusion_matrix = confusion_matrix(y_test, balanced_random_forest_predictions)
balanced_random_forest_confusion_matrix_df = pd.DataFrame(
    balanced_random_forest_confusion_matrix, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
)
balanced_random_forest_confusion_matrix_df

Predicted 0  Predicted 1
Actual 0           66           35
Actual 1         2298        14806

In [41]:
# Print the imbalanced classification report
print(classification_report(y_test, balanced_random_forest_predictions))

              precision    recall  f1-score   support

   high_risk       0.03      0.65      0.05       101
    low_risk       1.00      0.87      0.93     17104

    accuracy                           0.86     17205
   macro avg       0.51      0.76      0.49     17205
weighted avg       0.99      0.86      0.92     17205



In [43]:
# List the features sorted in descending order by feature importance
importances = balanced_random_forest_model.feature_importances_
importances_sorted = sorted(zip(balanced_random_forest_model.feature_importances_, X.columns), reverse=True)
importances_sorted[:10]

[(0.07492034883336385, 'total_pymnt_inv'),
 (0.06702915670653609, 'total_rec_prncp'),
 (0.057376824588817474, 'last_pymnt_amnt'),
 (0.056947623731881426, 'total_rec_int'),
 (0.0518570150166118, 'total_pymnt'),
 (0.03144189938268305, 'issue_d'),
 (0.028782832340719103, 'int_rate'),
 (0.01893502092613399, 'annual_inc'),
 (0.017793642816552515, 'mths_since_recent_inq'),
 (0.016906782057484433, 'out_prncp')]

### Easy Ensemble Classifier

In [44]:
# Train the Classifier
from imblearn.ensemble import EasyEnsembleClassifier
easy_ensemble_model = EasyEnsembleClassifier(n_estimators=100, random_state=78)
easy_ensemble_model = easy_ensemble_model.fit(X_train_scaled, y_train)
easy_ensemble_predictions = easy_ensemble_model.predict(X_test_scaled)

In [45]:
# Calculated the balanced accuracy score
easy_ensemble_accuracy_score = accuracy_score(y_test, easy_ensemble_predictions)
easy_ensemble_accuracy_score

0.9475733798314443

In [46]:
# Display the confusion matrix
easy_ensemble_confusion_matrix = confusion_matrix(y_test, easy_ensemble_predictions)
easy_ensemble_confusion_matrix_df = pd.DataFrame(
    easy_ensemble_confusion_matrix, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
)
easy_ensemble_confusion_matrix_df

Predicted 0  Predicted 1
Actual 0           92            9
Actual 1          893        16211

In [47]:
# Print the imbalanced classification report
print(classification_report(y_test, easy_ensemble_predictions))

              precision    recall  f1-score   support

   high_risk       0.09      0.91      0.17       101
    low_risk       1.00      0.95      0.97     17104

    accuracy                           0.95     17205
   macro avg       0.55      0.93      0.57     17205
weighted avg       0.99      0.95      0.97     17205



### Final Questions

1. Which model had the best balanced accuracy score?

The Easy Ensemble Classifier had the best balanced accuracy score (0.95), which was better than the Balanced Random Forest Classifier (0.86).

2. Which model had the best recall score?

The Easy Ensemble Classifier also had the best recall scores with a weighted average of 0.95 compared to the Balanced Random Forest Classifier which returned a 0.86 weighted average recall score.

3. Which model had the best geometric mean score?

The Easy Ensemble Classifier also had the best geometric mean scores with a weighted average of 0.97 compared to the Balanced Random Forest Classifier which returned a 0.92 weighted average geometric mean score.

4. What are the top three features?

The top 3 features are total_pymnt_inv, total_rec_prncp, and last_pymnt_amnt.